### PR TITLE
Added Postcopy migration test cases

### PR DESCRIPTION
**Pull request by balamuruhans**
Added postcopy migration scenarios
1. --postcopy-after-precopy : starts postcopy automatically after precopy migration
2.  --postcopy which starts precopy migration and have to trigger postcopy with migrate-postcopy

### CODE REVIEW

File: libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py [hunk 1]
```diff
@@ -63,16 +63,39 @@ def cleanup_dest(vm, src_uri=""):
...
+                successful = process.SubProcess(command, shell=True)
+                postcopy_pid = successful.start()
+                logging.debug("Postcopy enabled: %s and its pid is %s" %
+                              (command, postcopy_pid))
+                command = "/usr/bin/virsh %s %s" % (postcopy, vm_name)
+                while successful.poll() is None:
+                    time.sleep(5)
+                    if vm.is_alive() and check == 0:
+                        logging.debug("triggering postcopy with %s" % command)
+                        process.system_output(command, verbose=True,
```
Comment (sathnaga) [inline libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py]: Output of this command not getting used further, system_output is required?

File: libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py [hunk 2]
```diff
@@ -63,16 +63,39 @@ def cleanup_dest(vm, src_uri=""):
...
+                                              shell=True)
+                        check = 1
```
Comment (sathnaga) [inline libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py]: Any check required for setting this?

File: libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py [hunk 3]
```diff
@@ -63,16 +63,39 @@ def cleanup_dest(vm, src_uri=""):
...
+            command += "%s --domain %s --desturi %s %s" % (options, vm_name,
+                                                           dest_uri, extra)
+            try:
+                successful = process.SubProcess(command, shell=True)
+                postcopy_pid = successful.start()
+                logging.debug("Postcopy enabled: %s and its pid is %s" %
+                              (command, postcopy_pid))
+                command = "/usr/bin/virsh %s %s" % (postcopy, vm_name)
+                while successful.poll() is None:
+                    time.sleep(5)
```
Comment (sathnaga) [inline libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py]: sleep can be avoided?

File: libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py [hunk 4]
```diff
@@ -63,16 +63,39 @@ def cleanup_dest(vm, src_uri=""):
...
+            command = "/usr/bin/virsh -c 'qemu:///system' migrate "
+            command += "%s --domain %s --desturi %s %s" % (options, vm_name,
+                                                           dest_uri, extra)
+            try:
+                successful = process.SubProcess(command, shell=True)
+                postcopy_pid = successful.start()
+                logging.debug("Postcopy enabled: %s and its pid is %s" %
+                              (command, postcopy_pid))
+                command = "/usr/bin/virsh %s %s" % (postcopy, vm_name)
+                while successful.poll() is None:
```
Comment (sathnaga) [inline libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py]: Need an additional check on timeout, what if the command hangs for ever?

File: libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py (hunk 2 above)
Comment (sathnaga) [inline libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py]: There are few additional cleanup after migration, in vm.migrate https://github.com/avocado-framework/avocado-vt/blob/master/virttest/libvirt_vm.py#L1678, I guess you might have to think about handling it here, or
ve an another suggestion,
How about improving vm.migrate or virsh.migrate itself to handle these kind of additional commands
using subprocess 
Pros: 
1. we can measure additionally time taken for each migration.
2. we can use for additional commands like migrate-postcopy
cons:
1. vm.migrate or virsh.migrate becomes bulky

File: libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py (hunk 1 above)
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py]: @sathnaga output of command "migrate-postcopy" is not required as it triggers the actual postcopy migration for migration command with --postcopy option enabled.

Exception is raised if command "migrate-postcopy" fails in system_output

File: libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py (hunk 2 above)
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py]: I used check to execute the command only once as while loop continues till migration ends and vm will be alive in --live

File: libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py (hunk 3 above)
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py]: @sathnaga it takes 3-5 seconds to start precopy migration after migration command is given and migrate-postcopy have to be used to trigger postcopy migration

File: libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py (hunk 4 above)
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py]: @sathnaga sure, I will use additional check for timeout

File: libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py (hunk 2 above)
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py]: @waynesun09 please review this PR and give us suggestion on the above discussion.

File: libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py (hunk 1 above)
Comment (dzhengfy) [inline libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py]: It is ok to use 'system_output' or 'system' to execute a command, I think.

File: libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py (hunk 3 above)
Comment (dzhengfy) [inline libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py]: Yes, it is 3-5 seconds to start precopy migration which is also my test result. So I think it is ok to sleep for this period.

**Comment by dzhengfy** [comment]
In one word, I think this PR is only to test option "--postcopy-after-precopy", " --postcopy" combined with "migrate-postcopy" can execute and no error message happens.
But if you want to test the real function, that means it really take effect on the guest migration, I guess some more codes need to be added. 

For example,
1. Command 'migrate-postcopy' is to make the guest on target host change to running state immediately after this command is issued. Without this command, the guest should need much more time to be running which is the difference between postcopy and precopy. So perhaps some stress should be loaded in the guest to make the migration cost more time..., in order for this command to take effect in deed.
2. Add new check points, like virt-viewer process connecting to the guest is still connected...

Regarding 1, I think it is better to have. But if you only want to test an idle guest, it is ok to leave as it is.
Regarding 2, I suggest you add it.

**Comment by balamuruhans** [comment]
@dzhengfy ,
Comment 1: I agree that real functionality of postcopy migration can be tested if guest is stressed to cost time, but thought of adding preliminary check to test whether migrate-postcopy triggers and executed with out any errors (we face issues in normal execution of migrate-postcopy). Can I add the real functionality test case after this PR.

Comment 2: virt-viewer process connected from source host will get disconnected after migrating guest to destination host as I have observed this behaviour using virsh console. Please correct me if I am wrong.

**Comment by dzhengfy** [comment]
@balamuruhans ,
Regarding comment1, I am ok with your idle guest test
Regarding comment2, only spice supports seamless migration, I guess you were using vnc. Please check your guest XML.

BTW,  for postcopy_after_precopy option,  it should be used combined with "--postcopy" otherwise it will not take effect although we may not notice its behaviour.  So it should be like this:
virsh_migrate_options = "--live --postcopy --postcopy-after-precopy"

**Comment by balamuruhans** [comment]
@dzhengfy,
Comment 1: okay
Comment 2: Is this the one you mentioned <graphics type='vnc' port='-1' autoport='yes'/> ?

Sure, I will correct - virsh_migrate_options = "--live --postcopy --postcopy-after-precopy"

**Comment by balamuruhans** [comment]
graphics type='vnc' port='-1' autoport='yes'

**Comment by dzhengfy** [comment]
@balamuruhans ,
Yes.  You may configure like below:
1. In guest XML, replace "<graphics type='vnc' port='-1' autoport='yes'>" with "<graphics type='spice' port='5900' autoport='no' keymap='en-us'>"
2. configure qemu.conf to enable 'spice_listen = "0.0.0.0"' and restart libvirtd

**Comment by balamuruhans** [comment]
@dzhengfy, Thank you
1. the tags are not visible to replace, please paste the option without tag '< >' so that it is visible.
2. I will change the qemu conf accordingly

**Comment by dzhengfy** [comment]
@balamuruhans , 
I also have a requirement to execute other command during migration executes. So I think this maybe a common need.
So I submit a PR (https://github.com/avocado-framework/avocado-vt/pull/645). Would you like to have a look at it? Welcome your comments. Thank you.

**Comment by balamuruhans** [comment]
@dzhengfy , cool ... Sure I will

**Comment by balamuruhans** [comment]
working on to use https://github.com/avocado-framework/avocado-vt/pull/645 for my PR

**Comment by balamuruhans** [comment]
Fixed issue faced - https://github.com/avocado-framework/avocado-vt/pull/675

**Comment by balamuruhans** [comment]
# avocado run --vt-type libvirt --vt-config /var/lib/libvirt/images/bala/avocado-fvt-wrapper/data/avocado-vt/backends/libvirt/cfg/migrate.cfg --vt-only-filter 'RHEL.7.2.ppc64le scsi qcow2' --vt-no-filter 'network user macvtap'
JOB ID     : 2839b62f884e36289ad391dc46fa12938a6adf2b
JOB LOG    : /var/lib/libvirt/images/bala/avocado-fvt-wrapper/results/job-2016-08-22T15.52-2839b62/job.log
TESTS      : 4
 (1/4) guest_migration.guest_import.qemu.qcow2.scsi.smp2.virtio_net.RHEL.7.2.ppc64le.powerkvm-qemu.unattended_install.import.import.default_install.aio_native: PASS (54.91 s)
 (2/4) guest_migration.postcopy_migration.qemu.qcow2.scsi.smp2.virtio_net.RHEL.7.2.ppc64le.powerkvm-libvirt.virsh.migrate.postcopy_migration.postcopy_after_precopy: PASS (181.92 s)
 (3/4) guest_migration.postcopy_migration.qemu.qcow2.scsi.smp2.virtio_net.RHEL.7.2.ppc64le.powerkvm-libvirt.virsh.migrate.postcopy_migration.postcopy: PASS (201.48 s)
 (4/4) guest_migration.guest_remove.qemu.qcow2.scsi.smp2.virtio_net.RHEL.7.2.ppc64le.powerkvm-libvirt.remove_guest.without_disk: PASS (1.01 s)
RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0
JOB HTML   : /var/lib/libvirt/images/bala/avocado-fvt-wrapper/results/job-2016-08-22T15.52-2839b62/html/results.html
TESTS TIME : 439.31 s

**Comment by balamuruhans** [comment]
@dzhengfy, spice graphics tests I will add in it my next PR, also spice graphics is not yet supported in Power currently.

**Comment by balamuruhans** [comment]
@sathnaga @dzhengfy @waynesun09 Please help me in merging this PR. I have rebased this PR for #848

**Comment by sathnaga** [comment]
ACKed.

**Comment by balamuruhans** [comment]
Tested this PR with https://github.com/avocado-framework/avocado-vt/pull/675

**Comment by waynesun09** [comment]
ACK and merge now, thanks

**Comment by balamuruhans** [comment]
Thank you @waynesun09 @dzhengfy @sathnaga

**Pull request merged**